### PR TITLE
[refactor] set attention implementation

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As per title, refactors attention implementation setting and makes it a public API. We should encourage users to `model.set_attn_implementation()` whenever they want to change it after loading the model, instead of setting config's private attr `model.config._attn_implementation="sdpa"`

After the clean-up, we will be calling attention implementation setter only once per pretrained model class, when init the module. Since `from_pretrained/from_config` at the end call `init`, we don't need to keep it as a `classmethod`. Also setting attention after init allows us to know which backbones support attn or do not, and might be useful of we want to early raise errors in the future versions.

Also, removed redundant flags for FA2/FA3. Realized that we can use one flag for both versions :)

### CODE REVIEW

File: src/transformers/models/dpt/modeling_dpt.py [hunk 1]
```diff
@@ -109,7 +109,7 @@ def __init__(self, config, feature_size=None):
         patch_size = patch_size if isinstance(patch_size, collections.abc.Iterable) else (patch_size, patch_size)
         num_patches = (image_size[1] // patch_size[1]) * (image_size[0] // patch_size[0])
 
-        self.backbone = load_backbone(config)
+        self.backbone = load_backbone(config.backbone_config)
```
Comment (zucchini-nlp) [inline src/transformers/models/dpt/modeling_dpt.py]: Need to verify is this is correct, prob cc @qubvel 

The fix was needed becauuse backbone doesn't support `SDPA` but the model does. So `config.attn_implementation` and `config.backbone..attn_implementation` are different

Comment (qubvel) [reply src/transformers/models/dpt/modeling_dpt.py]: hmm, that's a breaking change in general, because load_backbone supports different ways of passing backbone params (other than just `backbone_config`). However, in case there is no way to avoid it and it doesn't break existing models on the hub - it should be ok

**Comment by zucchini-nlp** [comment]
Failing tests are unrelated, ready for review

**Comment by zucchini-nlp** [comment]
run-slow: bark, blip_2, instructblipvideo, modernbert, qwen2_5_vl, qwen2_vl, zamba

**Comment by zucchini-nlp** [comment]
+ removed redundant flags for FA2/FA3. Realized that we can use one flag for both versions :)

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -2259,81 +2251,64 @@ def _from_config(cls, config, **kwargs):
...
-                    )
+            try:
+                kernel = get_kernel(repo_id)
+                ALL_ATTENTION_FUNCTIONS.register(f"kernel_{repo_id.replace('/', '_')}", getattr(kernel, kernel_name))
+                attn_implementation = f"kernel_{repo_id.replace('/', '_')}"
+            except FileNotFoundError as e:
+                logger.warning(
+                    f"Could not find a kernel repository '{repo_id}' compatible with your devicein the hub: {e}. Using eager attention implementation instead."
+                )
+                attn_implementation = "eager"
```
Comment (Cyrilvallez) [inline src/transformers/modeling_utils.py]: I know this matches current behavior, but we should probably change it to sdpa to match default behavior, WDYT?

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -2465,64 +2416,57 @@ def can_generate(cls) -> bool:
...
         return False
 
-    @classmethod
-    def _check_and_enable_flash_attn_2(
-        cls,
-        config,
+    def _flash_attn_2_can_dispatch(
+        self,
         torch_dtype: Optional[torch.dtype] = None,
         device_map: Optional[Union[str, dict[str, int]]] = None,
```
Comment (Cyrilvallez) [inline src/transformers/modeling_utils.py]: Let's not have them as argument to the different functions, as we use the internal attrs anyway! (same comment for fa3)

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -2465,64 +2416,57 @@ def can_generate(cls) -> bool:
...
+                        )
+                    else:
+                        raise ImportError(f"{preface} Flash Attention 2 is not available. {install_message}")
+                elif torch.version.hip:
+                    if flash_attention_version < version.parse("2.0.4"):
+                        raise ImportError(
+                            f"{preface} you need flash_attn package version to be greater or equal than 2.0.4. Detected version {flash_attention_version}. {install_message}"
+                        )
+                    else:
+                        raise ImportError(f"{preface} Flash Attention 2 is not available. {install_message}")
```
Comment (Cyrilvallez) [inline src/transformers/modeling_utils.py:2466]: Don't we have a simple `is_fa2_installed` somewhere for all this? (same comment for fa3)

File: src/transformers/models/qwen2_vl/modeling_qwen2_vl.py [hunk 5]
```diff
@@ -1227,8 +1226,10 @@ def forward(
             )
             if attention_mask_tensor is not None and attention_mask_tensor.ndim == 4:
                 attention_mask_tensor = torch.diagonal(attention_mask_tensor[:, 0], dim1=1, dim2=2)
-                attention_mask_tensor = attention_mask_tensor / torch.finfo(attention_mask_tensor.dtype).min
-                attention_mask_tensor = (1.0 - attention_mask_tensor).int()
+                # Invert if floating, some attention interfaces pass already a boolean 4D mask
+                if attention_mask_tensor.is_floating_point():
+                    attention_mask_tensor = attention_mask_tensor / torch.finfo(attention_mask_tensor.dtype).min
+                    attention_mask_tensor = (1.0 - attention_mask_tensor).int()
```
Comment (Cyrilvallez) [inline src/transformers/models/qwen2_vl/modeling_qwen2_vl.py]: Looks nasty indeed, but unrelated right?

File: tests/generation/test_utils.py [hunk 6]
```diff
@@ -2007,7 +2015,7 @@ def test_generate_with_static_cache(self):
             max_new_tokens = 20
 
             for dtype in (torch.float32, torch.float16):
-                model = model_class(config).to(torch_device).to(dtype).eval()
+                model = model_class(copy.deepcopy(config)).to(torch_device).to(dtype).eval()
```
Comment (Cyrilvallez) [inline tests/generation/test_utils.py:2018]: Can you clarify why we need to start deepcopy all the configs? 🤗

**Review by Cyrilvallez — COMMENTED** [review]
Extremely nice and welcome PR! Super glad to see this become a public API, and to simplify how it's done overall! 🤗🚀

File: src/transformers/models/qwen2_vl/modeling_qwen2_vl.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/models/qwen2_vl/modeling_qwen2_vl.py]: It is related unfortunately. The attention tensor at some point with SDPA came in as boolean 4D mask, while in eager mode it is a floating point mask. AFAIK we support both types of masks from users

This part tries to revert ops back and get a 2D boolean mask, if a 4D mask is found. The 2D mask is later used by Qwen's special 3D position ids constructor. Actually it is same as https://github.com/huggingface/transformers/pull/39333, just realized

File: tests/generation/test_utils.py (hunk 6 above)
Comment (zucchini-nlp) [reply tests/generation/test_utils.py:2018]: Oh yeah, it was needed because when we do `Model(config)` the config changes its attention implementation in-place. It was like that always tbh, and I added `deepcopy` at some point when trying to remove `config._attn_implementation_autoset` attribute

Right now I reverted it back, it caused more problems than I thought. We could revert `deepcopy` as well, but I think it's more robust to keep it and ensure no `in-place` changes in original config. Another option is to use only `Model.from_config()` in tests, it deepcopies config internally

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/modeling_utils.py:2466]: Huh, indeed, ig we are checking the exact issue here why `not is_flash_attn_2_available()` to raise a proper and informative error. The import checker utils like `is_flash_attn_2_available()` usually return a simple boolean, and raise no errors

We try to raise informative errors in modeling code only, so I don't think we need to move this helper in `import_utilis`. WDYT?

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/modeling_utils.py]: makes sense, can do the usual "sdpa if available or fallback eager if not"

**Comment by zucchini-nlp** [comment]
We also need to make proper documentation about attentions, I am stuck at finding the right place. Existing docs are a bit dispersed and don't explain much how the API works internally :(

Let's leave it for next PR, might need to change doc structure slightly

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -2466,64 +2417,53 @@ def can_generate(cls) -> bool:
...
         Checks the availability of Flash Attention 2 and compatibility with the current model.
 
         If all checks pass and `hard_check_only` is False, the method will set the config attribute `attn_implementation` to "flash_attention_2" so that the model can initialize the correct attention module.
         """
-        if not cls._supports_flash_attn_2:
+        # Config always has `torch_dtype` but we need the next line for `no_super_init()` tests
+        torch_dtype = self.config.torch_dtype if hasattr(self.config, "torch_dtype") else torch.get_default_dtype()
+        device_map = self.hf_device_map if hasattr(self, "hf_device_map") else None
+
+        if not self._supports_flash_attn:
```
Comment (qubvel) [inline src/transformers/modeling_utils.py]: just curious if we should check for `_supports_flash_attn_2` attribute for BC with remote code models on the Hub

Comment (zucchini-nlp) [reply src/transformers/modeling_utils.py]: good point!

File: src/transformers/models/qwen2_vl/modeling_qwen2_vl.py (hunk 5 above)
Comment (Cyrilvallez) [reply src/transformers/models/qwen2_vl/modeling_qwen2_vl.py]: I see but what I meant is it's still an issue independently of this refactor! Fine to fix it here though no worries!

File: src/transformers/modeling_utils.py (hunk 4 above)
Comment (Cyrilvallez) [reply src/transformers/modeling_utils.py:2466]: Alright, we can keep as-is, at least for now! This is already a big PR!

**Review by Cyrilvallez — APPROVED** [review]
Alright, let's merge it!! 🔥🔥

**Comment by Cyrilvallez** [comment]
(But indeed nice documentation about this seem super important for users, let's keep it in mind somewhere!)

**Comment by zucchini-nlp** [comment]
Yep, doc coming in the next PR

**Review by ArthurZucker — COMMENTED** [review]
Nice!

**Comment by rangehow** [comment]
This PR does not seem to work as expected. It breaks the default attention mechanism of ModernBERT (it should be fa2 when the user does not specify it, but now it is sdpa). Moreover, this function is not actually called when the model is constructed through class initialization.

@zucchini-nlp @ArthurZucker
